File: PC4/client-desktop/src/main/java/dogmsg/client/LocalCache.java
package dogmsg.client;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;

/**
 * T-B8: Cache local de mensajes en SQLite.
 *
 * <p>Permite abrir una conversacion mostrando los mensajes recientes sin
 * consultar al servidor, y sincronizar cuando hay diferencias. Usa el driver
 * {@code sqlite-jdbc} (declarado en build.gradle). El acceso esta sincronizado
 * porque el hilo de UI y el de red pueden escribir a la vez.</p>
 */
public class LocalCache implements AutoCloseable {

    /** Representa una fila de mensaje cacheado. */
    public static class CachedMessage {
        public long id;
        public long peerId;       // usuario o grupo de la conversacion
        public boolean isGroup;
        public long senderId;
        public String type;       // text | image | file
        public String text;       // texto en claro (ya descifrado localmente)
        public long timestamp;
        public String status;     // sent | delivered | read
    }

    private final Connection conn;
    private final Object lock = new Object();

    public LocalCache(String dbPath) throws SQLException {
        this.conn = DriverManager.getConnection("jdbc:sqlite:" + dbPath);
        initSchema();
    }

    private void initSchema() throws SQLException {
        synchronized (lock) {
            try (Statement st = conn.createStatement()) {
                st.executeUpdate(
                    "CREATE TABLE IF NOT EXISTS cache_messages (" +
                    "  id INTEGER PRIMARY KEY AUTOINCREMENT," +
                    "  server_id INTEGER," +              // id en el servidor (si se conoce)
                    "  peer_id INTEGER NOT NULL," +       // contraparte de la conversacion
                    "  is_group INTEGER DEFAULT 0," +
                    "  sender_id INTEGER NOT NULL," +
                    "  type TEXT DEFAULT 'text'," +
                    "  text TEXT," +                      // texto en claro (descifrado)
                    "  timestamp INTEGER NOT NULL," +
                    "  status TEXT DEFAULT 'sent'" +
                    ")");
                st.executeUpdate(
                    "CREATE INDEX IF NOT EXISTS idx_peer_time " +
                    "ON cache_messages(peer_id, is_group, timestamp)");
            }
        }
    }

    /** Inserta un mensaje en el cache local y devuelve su id local. */
    public long insert(CachedMessage m) throws SQLException {
        synchronized (lock) {
            String sql = "INSERT INTO cache_messages " +
                    "(server_id, peer_id, is_group, sender_id, type, text, timestamp, status) " +
                    "VALUES (?,?,?,?,?,?,?,?)";
            try (PreparedStatement ps = conn.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
                ps.setLong(1, m.id);
                ps.setLong(2, m.peerId);
                ps.setInt(3, m.isGroup ? 1 : 0);
                ps.setLong(4, m.senderId);
                ps.setString(5, m.type);
                ps.setString(6, m.text);
                ps.setLong(7, m.timestamp);
                ps.setString(8, m.status);
                ps.executeUpdate();
                try (ResultSet keys = ps.getGeneratedKeys()) {
                    return keys.next() ? keys.getLong(1) : -1;
                }
            }
        }
    }

    /** Carga los ultimos {@code limit} mensajes de una conversacion (cronologico). */
    public List<CachedMessage> recent(long peerId, boolean isGroup, int limit) throws SQLException {
        synchronized (lock) {
            String sql = "SELECT id, peer_id, is_group, sender_id, type, text, timestamp, status " +
                    "FROM cache_messages WHERE peer_id=? AND is_group=? " +
                    "ORDER BY timestamp DESC LIMIT ?";
            try (PreparedStatement ps = conn.prepareStatement(sql)) {
                ps.setLong(1, peerId);
                ps.setInt(2, isGroup ? 1 : 0);
                ps.setInt(3, limit);
                List<CachedMessage> out = new ArrayList<>();
                try (ResultSet rs = ps.executeQuery()) {
                    while (rs.next()) {
                        CachedMessage m = new CachedMessage();
                        m.id = rs.getLong("id");
                        m.peerId = rs.getLong("peer_id");
                        m.isGroup = rs.getInt("is_group") == 1;
                        m.senderId = rs.getLong("sender_id");
                        m.type = rs.getString("type");
                        m.text = rs.getString("text");
                        m.timestamp = rs.getLong("timestamp");
                        m.status = rs.getString("status");
                        out.add(m);
                    }
                }
                java.util.Collections.reverse(out); // a orden cronologico ascendente
                return out;
            }
        }
    }

    /** Actualiza el estado de un mensaje (p.ej. sent -> delivered). */
    public void updateStatus(long localId, String status) throws SQLException {
        synchronized (lock) {
            try (PreparedStatement ps = conn.prepareStatement(
                    "UPDATE cache_messages SET status=? WHERE id=?")) {
                ps.setString(1, status);
                ps.setLong(2, localId);
                ps.executeUpdate();
            }
        }
    }

    @Override
    public void close() {
        try { conn.close(); } catch (SQLException ignored) {}
    }
}

File: PC4/client-android/app/src/main/java/com/dogmsg/android/LocalCache.kt
package com.dogmsg.android

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper

/**
 * T-B8 (lado Android): cache local de mensajes en SQLite.
 *
 * Mismo esquema que LocalCache.java del cliente desktop, pero usando el
 * SQLiteOpenHelper nativo de Android (sin drivers externos). Permite abrir una
 * conversacion mostrando los mensajes recientes sin consultar al servidor.
 */
class LocalCache(context: Context) :
    SQLiteOpenHelper(context, DB_NAME, null, DB_VERSION) {

    /** Representa una fila de mensaje cacheado. */
    data class CachedMessage(
        var id: Long = 0,
        var peerId: Long = 0,        // usuario o grupo de la conversacion
        var isGroup: Boolean = false,
        var senderId: Long = 0,
        var type: String = "text",   // text | image | file
        var text: String? = null,    // texto en claro (ya descifrado localmente)
        var timestamp: Long = 0,
        var status: String = "sent"  // sent | delivered | read
    )

    /** Resumen de una conversacion: ultimo mensaje, para poblar la lista de chats. */
    data class ConversationSummary(
        val peerId: Long,
        val isGroup: Boolean,
        val lastText: String?,
        val lastType: String,
        val lastTimestamp: Long
    )

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL(
            """
            CREATE TABLE IF NOT EXISTS cache_messages (
              id INTEGER PRIMARY KEY AUTOINCREMENT,
              server_id INTEGER,
              peer_id INTEGER NOT NULL,
              is_group INTEGER DEFAULT 0,
              sender_id INTEGER NOT NULL,
              type TEXT DEFAULT 'text',
              text TEXT,
              timestamp INTEGER NOT NULL,
              status TEXT DEFAULT 'sent'
            )
            """.trimIndent()
        )
        db.execSQL(
            "CREATE INDEX IF NOT EXISTS idx_peer_time " +
                    "ON cache_messages(peer_id, is_group, timestamp)"
        )
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        db.execSQL("DROP TABLE IF EXISTS cache_messages")
        onCreate(db)
    }

    /** Inserta un mensaje en el cache local y devuelve su id local. */
    @Synchronized
    fun insert(m: CachedMessage): Long {
        val values = ContentValues().apply {
            put("server_id", m.id)
            put("peer_id", m.peerId)
            put("is_group", if (m.isGroup) 1 else 0)
            put("sender_id", m.senderId)
            put("type", m.type)
            put("text", m.text)
            put("timestamp", m.timestamp)
            put("status", m.status)
        }
        return writableDatabase.insert("cache_messages", null, values)
    }

    /** Carga los ultimos [limit] mensajes de una conversacion (orden cronologico). */
    @Synchronized
    fun recent(peerId: Long, isGroup: Boolean, limit: Int): List<CachedMessage> {
        val out = ArrayList<CachedMessage>()
        val cursor = readableDatabase.rawQuery(
            "SELECT id, peer_id, is_group, sender_id, type, text, timestamp, status " +
                    "FROM cache_messages WHERE peer_id=? AND is_group=? " +
                    "ORDER BY timestamp DESC LIMIT ?",
            arrayOf(peerId.toString(), (if (isGroup) 1 else 0).toString(), limit.toString())
        )
        cursor.use { c ->
            while (c.moveToNext()) {
                out.add(
                    CachedMessage(
                        id = c.getLong(0),
                        peerId = c.getLong(1),
                        isGroup = c.getInt(2) == 1,
                        senderId = c.getLong(3),
                        type = c.getString(4) ?: "text",
                        text = c.getString(5),
                        timestamp = c.getLong(6),
                        status = c.getString(7) ?: "sent"
                    )
                )
            }
        }
        out.reverse() // a orden cronologico ascendente
        return out
    }

    /** Actualiza el estado de un mensaje (p.ej. sent -> delivered). */
    @Synchronized
    fun updateStatus(localId: Long, status: String) {
        val values = ContentValues().apply { put("status", status) }
        writableDatabase.update("cache_messages", values, "id=?", arrayOf(localId.toString()))
    }

    /**
     * Devuelve el ultimo mensaje de cada conversacion distinta (peer_id +
     * is_group), ordenado por mas reciente primero. Se usa para poblar la
     * lista de chats al abrir la app o al volver de un ChatActivity --
     * antes esa lista solo se llenaba en memoria via upsert() en vivo, asi
     * que se vaciaba (aparentando "no tener chats") cada vez que el proceso
     * se reciclaba o MainActivity se recreaba, aunque los mensajes seguian
     * persistidos aqui en SQLite.
     */
    @Synchronized
    fun conversationSummaries(): List<ConversationSummary> {
        val out = ArrayList<ConversationSummary>()
        val cursor = readableDatabase.rawQuery(
            "SELECT m.peer_id, m.is_group, m.text, m.type, m.timestamp " +
                    "FROM cache_messages m " +
                    "INNER JOIN (" +
                    "  SELECT peer_id, is_group, MAX(timestamp) AS max_ts " +
                    "  FROM cache_messages GROUP BY peer_id, is_group" +
                    ") last ON m.peer_id = last.peer_id AND m.is_group = last.is_group " +
                    "  AND m.timestamp = last.max_ts " +
                    "ORDER BY m.timestamp DESC",
            null
        )
        cursor.use { c ->
            while (c.moveToNext()) {
                out.add(
                    ConversationSummary(
                        peerId = c.getLong(0),
                        isGroup = c.getInt(1) == 1,
                        lastText = c.getString(2),
                        lastType = c.getString(3) ?: "text",
                        lastTimestamp = c.getLong(4)
                    )
                )
            }
        }
        return out
    }

    companion object {
        private const val DB_NAME = "dogmsg_cache.db"
        private const val DB_VERSION = 1
    }
}